contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,620 @@
-   server or process only the information they receive from the default scheduler.
+Many features are being added to the Kubernetes Scheduler. They keep making the
+code larger and the logic more complex. A more complex scheduler is harder to
+maintain, its bugs are harder to find and fix, and those users running a custom
+scheduler have a hard time catching up and integrating new changes. The current
+Kubernetes scheduler provides [webhooks to extend](./scheduler_extender.md) its
+functionality. However, these are limited in a few ways:
+
+1.  The number of extension points are limited: "Filter" extenders are called
+    after default predicate functions. "Prioritize" extenders are called after
+    default priority functions. "Preempt" extenders are called after running
+    default preemption mechanism. "Bind" verb of the extenders are used to bind
+    a Pod. Only one of the extenders can be a binding extender, and that
+    extender performs binding instead of the scheduler. Extenders cannot be
+    invoked at other points, for example, they cannot be called before running
+    predicate functions.
+1.  Every call to the extenders involves marshaling and unmarshalling JSON.
+    Calling a webhook (HTTP request) is also slower than calling native
+    functions.
+1.  It is hard to inform an extender that scheduler has aborted scheduling of a
+    Pod. For example, if an extender provisions a cluster resource and scheduler
+    contacts the extender and asks it to provision an instance of the resource
+    for the pod being scheduled and then scheduler faces errors scheduling the
+    pod and decides to abort the scheduling, it will be hard to communicate the
+    error with the extender and ask it to undo the provisioning of the resource.
+1.  Since current extenders run as a separate process, they cannot use
+    scheduler's cache. They must either build their own cache from the API
+    server or process only the information they receive from the default
+    scheduler.
 
 The above limitations hinder building high performance and versatile scheduler
-extensions. We would ideally like to have an extension mechanism that is fast
-enough to allow keeping a bare minimum logic in the scheduler core and convert
-many of the existing features of default scheduler, such as predicate and
-priority functions and preemption into plugins. Such plugins will be compiled
-with the scheduler. We would also like to provide an extension mechanism that do
-not need recompilation of scheduler. The expected performance of such plugins is
-lower than in-process plugins. Such out-of-process plugins should be used in
-cases where quick invocation of the plugin is not a constraint.
+features. We would ideally like to have an extension mechanism that is fast
+enough to allow existing features to be converted into plugins, such as
+predicate and priority functions. Such plugins will be compiled into the
+scheduler binary.
```

Comment:
If I'm reading this correctly, I think this is a step back from the current approach as it creates a tight coupling between kube-scheduler and the plugin.

Currently, there is no dependency between the plugin and the version of the kube-scheduler. You can update your kubernetes cluster (ie move to a newer version of kube-scheduler) and not have to update the plugin since the interface between them remains the same.

This approach would require anyone who has implemented a plugin to recompile and publish their plugin+kube-scheduler image with each release of kubernetes, which is an additional burden on the maintainer.

Reply:
You are not wrong. With extenders today, you update to a new scheduler version by modifying the manifest for your custom scheduler, and you do not need to build any new images.

The existing extender interface is not being removed as part of this framework (and I don't think we plan to deprecate it). So authors can choose between two options. (A) loose coupling, easy upgrades, language choice, or (B) tight coupling, rebuild to upgrade, high performance.

Also, we are juggling two classes of use case here. We want to migrate existing features of the standard scheduler to plugins so that our code is easier to maintain & debug. There are also many people who want to write custom schedulers.

@bsalamat Do you have thoughts on this topic?

Reply:
This approach IMO doesnt take away the old way of writing custom schedulers without reusing any of the core scheduler code and hence not needing to vendor that. So this essentially adds one more option to write a custom scheduler which can reuse features of the existing scheduler and people may choose to not go this route if they would rather prefer the extender approach or the custom scheduler without the tight coupling approach